oeps/oep-0052-arch-event-bus-technology.rst

```diff
@@ -0,0 +1,264 @@
+
+.. _Amazon MSK: https://aws.amazon.com/msk/
+.. _Confluent Platform: https://www.confluent.io/product/confluent-platform
+.. _comparison of Apache Kafka vs Vendors: https://www.kai-waehner.de/blog/2021/04/20/comparison-open-source-apache-kafka-vs-confluent-cloudera-red-hat-amazon-msk-cloud/
+
+Kafka Highlights
+~~~~~~~~~~~~~~~~
+
+Pros
+^^^^
+
+* Battle-tested, widely adopted, big community, lots of documentation and answers.
+* Amazon MSK (AWS service) provides hosted path of least resistance.
+* `New Relic integration with Amazon MSK`_ (useful to edX.org).
+
+Cons
+^^^^
+
+* Many open questions about add-ons required for developers and operators.
+* Complex to manage, including likely manual scaling.
+
+.. _New Relic integration with Amazon MSK: https://docs.newrelic.com/docs/integrations/amazon-integrations/aws-integrations-list/aws-managed-kafka-msk-integration/
+
+Consequences
+------------
+
+* Operators will need to deploy and manage the selected infrastructure, which is likely to be complex. If Apache Kafka is selected, there are likely to be a set of auxiliary parts to provide all required functionality for our message bus.
+* Education will be required for both developers and operators regarding best practices for each role.
+* Code to interact with Kafka and its libraries will be added to core services.
+* At least one initial use case must be completed. One potential candidate is the grade change event in the LMS, and its use by the Credentials service.
+* Once we have a message bus, we can investigate other potential use cases:
+
+  * Course/program update propagation.
+  * Feed into xAPI/Caliper capabilities.
+  * New services and features can be built fully de-coupled from the core application.
+
+Rejected Alternatives
+---------------------
+
+Apache Pulsar
+~~~~~~~~~~~~~
+
+Although rejected to start, `Apache Pulsar`_ remains an option if solving with Kafka turns out to be overly burdensome for developers or operators.
+
+Pros
+^^^^
+
+* Ease of scalability (built-in, according to docs).
+* Ease of data retention capabilities.
+* Additional built-in pub/sub features (built-in, according to docs).
+
+Cons
+^^^^
+
+* Requires 3rd party hosting or larger upfront investment in self-hosted (kubernetes).
+* Less mature (but growing) community, little documentation, and few answers.
+
+Note: Read an interesting (Kafka/Confluent) biased article exploring `comparisons and myths of Kafka vs Pulsar`_.
+
+.. _Apache Pulsar: https://pulsar.apache.org/
+.. _comparisons and myths of Kafka vs Pulsar: https://dzone.com/articles/pulsar-vs-kafka-comparison-and-myths-explored
+
+Redis
+~~~~~
+
+Pros
+^^^^
+
+* Already part of Open edX platform
+
+Cons
+^^^^
+
+* Can lose acked data, even if RAM backed up with an append-only file (AOF).
+* Requires homegrown schema management.
```

Comment:
Isn't it also the case of Kafka? This is the reason why stream producers use an encoding format such as Protocol Buffers or Avro.

Reply:
Not entirely. Both Pulsar (natively) and Kafka (as a Confluent created add-on) have support for schema enforcement in the write path.

Reply:
The features brought by native schema management seem minimal compared to bringing our own data serialization format to the table. Also, the Pulsar Python client supports only a subset of the schema features: https://docs.google.com/spreadsheets/d/1iTzn6QnOpTYK8QQHxXRb6a25nvuTNGPPJDdUQHGBUS4/edit#gid=1784579914